state/src/applicator/task_queue.rs

```diff
@@ -128,6 +133,18 @@ impl StateApplicator {
             .get_queue_key_for_task(&task_id)?
             .ok_or_else(|| StateApplicatorError::Rejected(invalid_task_id(task_id)))?;
 
+        // If the task being popped is not at the top of the queue, reject the
+        // transition
+        let queued_tasks = tx.get_queued_tasks(&key)?;
+        let top_of_queue = queued_tasks.first();
+        match top_of_queue {
+            None => return Err(StateApplicatorError::Rejected(task_not_in_queue(task_id, key))),
+            Some(task) if task.id != task_id => {
+                return Err(StateApplicatorError::Rejected(task_not_in_queue(task_id, key)));
+            },
+            _ => {},
+        }
```

Comment:
nit: up to preference but could rewrite this as

```rust
if top_of_queue.is_none() || top_of_queue.unwap().id != task_id {
    return ...;
}
```

slightly more readable IMO